src/backend/distributed/executor/executor.c

```diff
@@ -1286,6 +1373,20 @@ ConnectionStateMachine(WorkerSession *session)
 
 			case MULTI_CONNECTION_CONNECTED:
 			{
+				/* if we're expanding the nodes in a transaction, use 2PC */
+				if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC &&
+					TransactionModifiedDistributedTable(execution) &&
+					DistributedPlanModifiesDatabase(execution->plan) &&
+					!ConnectionModifiedPlacement(session->connection))
+				{
+					/*
+					 * We already did a modification, but not on the connection that we
+					 * just opened, which means we're now going to make modifications
+					 * over multiple connections. Activate 2PC!
+					 */
+					CoordinatedTransactionUse2PC();
```

Comment:
Why is the check in `StartDistributedExecution()` for activating 2PC not enough? What are some examples? Is this for modify tasks with single shard but multiple placements? Otherwise I guess both `DistributedStatementRequiresRollback()` and `TaskListRequires2PC()` will return `true` if the checks here are `true` and we have already enabled 2PC in `StartDistributedExecution()`?

In any case, some comments on why this is necessary might be useful.